homework-submission/week1/src/App.js

```diff
@@ -0,0 +1,65 @@
+import React, { Component } from 'react';
+import logo from './logo.svg';
+import './App.css';
+import Todos from './components/Todos';
+import AddTodo from './components/AddTodo';
+import data from './sources/data.json'
+
+class App extends Component {
+  state = {
+    data
+  }
+    handler = e => {
+      let todos = this.state.data;
+      todos[e.target.id -1].done = !todos[e.target.id -1].done;
```

Comment:
This way you are mutating the todos **directly** which is the behaviour that React is trying to avoid!.

Taking the `todos` in a variables like this `let todos = this.state.data;` doesn't create a new copy from `data` but instead it creates a new (pointer) to the same place where we store `data`. Based on this, line 14 will be updating the state **directly**.

What we need to do is to take a copy from our state which is **data** and update this copy without touching the original copy that we have in our state. Then, we ask the `this.setState` to update the state for us, which is the intended behaviour from React!.

There're many ways to take a copy from an object or array, one of them is `let todos = [...this.state.data]` (es6 syntax) and another way is `let todos = this.state.data.map( (item) => item);` and `let todos = this.state.data.slice()` and many other ways that you search about :) ..

However, In this case it is better to . use **filter**

```
this.state.data.map( (todo, i) => (
   {...todo, done: (e.target.id -1 === i) ? !todo.done : todo.done }
))
```

Please if this not clear, don't hesitate to contact me on slack to clarify it more :)